main/file.c

```diff
@@ -213,52 +237,150 @@ static void on_send(sub_send_cb_data_t *send_data){
 
         if(current_time_ms - sub_struct.last_tx_time_ms > sub_struct.tx_nack_timeout_ms){
 
-            if(xSemaphoreTake(sub_struct.mutex, portMAX_DELAY) == pdTRUE){
-
-                sub_struct.interval_ms = sub_struct.transmit_periods[SLEEP_MS];
-
-                xSemaphoreGive(sub_struct.mutex);
-
-            }
+            set_interval_ms(sub_struct.transmit_periods[SLEEP_MS]);
         }
     }
 }
 
 static void on_recive(sub_recv_cb_data_t *recv_data){
 
-    uint8_t obc_mac_addr[ESP_NOW_ETH_ALEN] = OBC_MAC_ADDRESS;
+    if(recv_data == NULL){
+        return;
+    }
+
+    uint8_t mcb_mac_addr[ESP_NOW_ETH_ALEN] = MCB_MAC_ADDRESS;
+    uint8_t broadcast_mac_addr[ESP_NOW_ETH_ALEN] = BROADCAST_MAC_ADDRESS;
 
-    if(memcmp(recv_data->mac, obc_mac_addr, ESP_NOW_ETH_ALEN)==0){
+    if(memcmp(recv_data->src_mac, mcb_mac_addr, ESP_NOW_ETH_ALEN)!=0){
 
-        if(recv_data->data_size == STATE_MSG_SIZE && recv_data->data[0] < ENUM_MAX){
+        return;
+    }
 
-            if(xSemaphoreTake(sub_struct.mutex, portMAX_DELAY) == pdTRUE){
+    if(memcmp(recv_data->dest_mac, broadcast_mac_addr, ESP_NOW_ETH_ALEN)==0){
 
-                sub_struct.interval_ms = sub_struct.transmit_periods[recv_data->data[0]];
+        if(recv_data->data_size == STATE_MSG_SIZE && recv_data->cmd.raw[0] < ENUM_MAX){
 
-                xSemaphoreGive(sub_struct.mutex);
-            }
+            sub_struct.status = recv_data->cmd.raw[0];
         }
     }
-    else{
-
+    else if(memcmp(recv_data->dest_mac, sub_struct.dev_mac_address, ESP_NOW_ETH_ALEN)==0){
         recv_cb_cmd_t temp_data;
 
-        memcpy(&temp_data, recv_data->data, recv_data->data_size);
+        memcpy(&temp_data, &recv_data->cmd, recv_data->data_size);
 
         if(sub_struct._on_data_rx == NULL){
 
             if(recv_data->data_size > MAX_DATA_LENGTH){
```

Comment:
Tego nie powinniśmy sprawdzać przed skopiowaniem danych? Bo tak to możemy wcześniej skopiować poza pamięć 😨

Reply:
a w zasadzie muszę tu w ogóle kopiować? nie mogę na kolejke wrzucić poprostu &recv_data->cmd?

Reply:
możesz